### DISCUSSION TITLE
v0.3.0 Testing Feedback

### DISCUSSION THREAD

**Discussion by tomvothecoder** [xCDAT/xcdat#296] [4. General]
You can post any feedback, questions, or comments in this post. GitHub discussions is threads based, so it will be easy to follow the individual conversations.

**UPDATE (8/31/22)**
Just an FYI that v0.3.1  was released which includes additional bug fixes that you might have come across in v0.3.0.
Check out the GitHub discussions post here: https://github.com/xCDAT/xcdat/discussions/322

**Comment by msahn** [comment]
## Feedback
I have tested xCDAT v0.3.0 following the [test instruction](https://docs.google.com/document/d/1IOGGgrr0qzI9IvPB0lmqpP0KD0OKPqkF4_BmKH0-M1o/edit#). A metric measuring precipitation variability across timescales ([code](https://github.com/PCMDI/pcmdi_metrics/tree/main/pcmdi_metrics/precip_variability), [demo](https://github.com/PCMDI/pcmdi_metrics/blob/main/doc/jupyter/Demo/Demo_7_precip_variability.ipynb)), one of the metrics implemented in [PMP](https://github.com/PCMDI/pcmdi_metrics), has been converted to use xCDAT instead of CDAT.

Below is the library script I modified to convert from CDAT to xCDAT. The APIs are changed for open&write nc files, horizontal_regridding, data concatenating, land-sea masking, spatial average, and others.
- [library using CDAT](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_cdat/code/lib_variability_across_timescales.py)
- [library using xCDAT](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_xcdat/code/lib_variability_across_timescales.py)

The differences in JSON results are negligibly small when I test it with demo data. 
- using CDAT: [demo notebook](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_cdat/Demo_7_precip_variability.ipynb), [JSON result](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_cdat/demo_output/precip_variability/GISS-E2-H/PS_pr.day_regrid.180x90_area.freq.mean_GISS-E2-H.historical.json)
- using xCDAT: [demo notebook](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_xcdat/Demo_7_precip_variability.ipynb), [JSON result](https://github.com/msahn/CDAT_to_xCDAT/blob/main/pcmdi_metrics/precip_variability_demo_xcdat/demo_output/precip_variability/GISS-E2-H/PS_pr.day_regrid.180x90_area.freq.mean_GISS-E2-H.historical.json)

## Question
While converting my script to use xCDAT, I noticed that most xCDAT APIs are based on "xarray.Dataset". Could it be possible to make it based on "xarray.DataArray" as well?

**Reply by tomvothecoder** [reply]
Thank you @msahn for posting your experimentation results here!

> While converting my script to use xCDAT, I noticed that most xCDAT APIs are based on "xarray.Dataset". Could it be possible to make it based on "xarray.DataArray" as well?

We spent a few months in the early development stage prototyping APIs with `xarray.DataArray` because it is similar to the `cdms2.TransientVariable`. However, we shifted to basing xCDAT APIs on `xarray.Dataset` because of limitations with the `DataArray` API (inability to hold bounds as a data variable). This is mentioned in the API Reference Overview here: https://xcdat.readthedocs.io/en/latest/api.html#overview

Here's helpful documentation on `xarray.Dataset`: https://docs.xarray.dev/en/stable/user-guide/data-structures.html#dataset

**Reply by mzelinka** [reply]
I just started playing around with xCDAT and had this same question.  If I want to compute a spatial average on a quantity that is not a dataset (because of previous operations that have changed it from a dataset to a dataarray), how do I do it? Is there a way to assign the spatial coordinate bounds to a dataarray, based on its parent dataset?

**Reply by pochedls** [reply]
It seems like this isn't possible (at least not now). Considerable time went into figuring out how to do this, including adding accessors to modify the xarray dataarray structure to hold bounds. It didn't work (I think the issue was that user added accessors are not persistent). Others have raised this issue (e.g., [here](https://github.com/pydata/xarray/issues/1475)). 

I think this is why xarray libraries sometimes pass the dataset around in function calls (even for a function that you think might be able to operate on just the dataarray, e.g., [here](https://xesmf.readthedocs.io/en/latest/notebooks/Rectilinear_grid.html#Perform-regridding))

**Reply by durack1** [reply]
@mzelinka did this example (of a subset region) work? https://xcdat.readthedocs.io/en/latest/examples/spatial-average.html#3.-Tropical-Region

**Reply by mzelinka** [reply]
yes, as long as I am working with a Dataset and not a DataArray it works fine.

**Comment by mzelinka** [comment]
I am having an issue reading in a netcdf file with xcdat, but xarray does not throw an error:
![Screen Shot 2022-08-23 at 2 22 33 PM](https://user-images.githubusercontent.com/11380489/186268725-97a2671d-b7b3-486e-9849-43a9ea0c0d1f.png)

**Reply by mzelinka** [reply]
I think the issue is with the empty "height" coordinate....

**Reply by pochedls** [reply]
Sorry about this, Mark. I believe [this issue](https://github.com/xCDAT/xcdat/issues/278) and a bunch of other issues were fixed in [version 0.3.1](https://github.com/xCDAT/xcdat/discussions/322). 

I think you can upgrade with `conda install -c conda-forge xcdat=0.3.1`

**Reply by mzelinka** [reply]
Thanks - that worked!

**Comment by mzelinka** [comment]
Ok, I have a new one (this may be related to my lack of understanding of xarray).  I have a Dataset from one experiment called 'pert' for which I can effectively compute the spatial average.  I have a Dataset from another control experiment called 'mocon' for which I can also effectively compute the spatial average. All good so far. When I difference these two Datasets to compute an anomaly Dataset called 'anom', it loses all the attributes. So I create a new Dataset 'anom2' that has the attributes from 'pert' assigned to 'anom'. But now when I compute a spatial average, only NaNs are returned. What am I doing wrong? See screenshots below:
![Screen Shot 2022-08-24 at 9 45 32 AM](https://user-images.githubusercontent.com/11380489/186475736-3c98dd43-4e49-49a2-b8c6-1a47a1638a3a.png)
![Screen Shot 2022-08-24 at 9 45 48 AM](https://user-images.githubusercontent.com/11380489/186475784-b0a80715-a89f-4789-b635-4f6d63602317.png)
![Screen Shot 2022-08-24 at 9 46 15 AM](https://user-images.githubusercontent.com/11380489/186475890-e3cdb4cb-4582-4ed0-b2a4-8bb4faf1373e.png)

**Reply by pochedls** [reply]
Thanks for catching this @mzelinka. I see why this is confusing. 

`anom = pert['tas'] - mocon` is subtracting two datasets from one another. When you do this `xarray` performs a subtraction operation on all dataarrays, including `lat_bnds` and `lon_bnds`. These arrays thus become `[[0., 0.][0., 0.], ..., [0., 0.]]`. These are what are used to calculate the spatial weights (the weights are thus zero, which ends up resulting in `NaN` values in the spatial average operation).

A way to do this:
```
# get anomaly dataarray
daa = ds2.tas - ds1.tas
# convert to a dataset
dsa = daa.to_dataset()
# the bounds were dropped - add them back
# note that ds1.lat.bounds returns the name of the bounds for lat in ds1 (i.e., a string)
# note that ds1.bounds.get_bounds('Y') gets the latitude bounds object (i.e., a dataarray)
dsa[ds1.lat.bounds] = ds1.bounds.get_bounds('Y')
dsa[ds1.lon.bounds] = ds1.bounds.get_bounds('X')
# do spatial average
dsaa = dsa.spatial.average('tas')
# check values
dsaa.tas.values
```

> array([0.50558825, 0.71024598, 0.64418223, ..., 6.66370024, 6.93816916,
>        7.04828337])

I wonder if it is worth adding an `xcdat.dataset.to_dataset()` function that ensures the bounds are included in the new dataset? Or another way to handle this kind of operation more easily.

**Reply by mzelinka** [reply]
Fascinating. Seems like xarray should know not to difference the data variables with '_bnds' in their name.  Seems like strange default behavior.

**Reply by tomvothecoder** [reply]
Here's a simpler way to do this, which stays on the Dataset level to maintain bounds and avoid the additional steps for creating a new anomalies Dataset.

```python
# 1. Calculate anomalies using the "tas" var from both datasets as new data variable, "tas_anom"
ds2["tas_anom"] = ds2.tas - ds1.tas

# 2. Perform spatial average on "tas_anom" 
ds2 = ds2.spatial.average("tas_anom")
```

> Fascinating. Seems like xarray should know not to difference the data variables with '_bnds' in their name. Seems like strange default behavior.

In a Dataset, xarray typically stores bounds as data variables and not coordinate variables. Arithmetic at the Dataset level applies to all data variables, including bounds stored as data variables.

To avoid this behavior, you have to perform arithmetic on specific data variables within the Dataset.

```python
# This performs arithmetic on all data variables, including bounds.
ds_new = ds2 - ds1

# This performs arithmetic only on `tas`.
ds_new["tas"] = ds2["tas"] - ds1["tas"]
```

**Comment by mzelinka** [comment]
I'm curious if the following behavior is expected / desired or if it is a bug.  If load in a dataset, the data_var I want has the name 'sst'.  But if I swap the longitudes using xcdat, this variable has a new name 'sea_surface_temperature'.   I'm using xcdat v0.3.3. See screen shot below:
![Screen Shot 2022-11-01 at 1 14 38 PM](https://user-images.githubusercontent.com/11380489/199332116-85398a5c-278f-44d1-a901-0e28cc74a451.png)

**Reply by pochedls** [reply]
This looks like a bug to me...

**Reply by durack1** [reply]
@mzelinka are the data arrays identical before and after this operation? So e.g. comparing a single timestep..

**Reply by tomvothecoder** [reply]
Thanks @mzelinka, I'll take a look into this.

**Reply by mzelinka** [reply]
@durack1 they look identical to me; just the variable name changed

**Reply by tomvothecoder** [reply]
I'm not exactly sure why this happens in v0.3.3, but it might have been a variable key reference.

This is fixed in #343 (as a side effect of refactoring), which is still in progress and to be included in v0.3.4.

**Reply by pochedls** [reply]
Just to follow-up, this changes in `xcdat.axis._align_to_360` when the [data variable associated with `lon` are fetched](https://github.com/xCDAT/xcdat/blob/31c0ad1f7afac8c62571cadeac569b56c072ea07/xcdat/axis.py#L295) via `ds.cf.data_vars.items()` which apparently uses `standard_name` as the key rather than the variable ID.

**Reply by tomvothecoder** [reply]
@pochedls Thank you for identifying the root cause. That might be the expected behavior with `cf_xarray`, since it interprets and maps CF attributes to variables. In hindsight, I should have used xarray's `ds.data_vars.items()` to reference the original keys. 
It is fixed though so we're good to go!

**Reply by mzelinka** [reply]
great! thanks @tomvothecoder and @pochedls !

**Reply by pochedls** [reply]
@mzelinka - I think this fix is in the new [xcdat release](https://github.com/xCDAT/xcdat/releases/tag/v0.4.0). It might take a bit a few hours to propagate to conda, though.

**Reply by tomvothecoder** [reply]
`xcdat=0.4.0` is now available on [conda-forge](https://anaconda.org/conda-forge/xcdat/files).

I just created a fresh environment with v0.4.0 and can confirm this is now fixed.

```python
import xcdat

print(xcdat.__version__)
# 0.4.0

ds = xcdat.open_dataset("/p/user_pub/climate_work/zelinka1/HadISST/HadISST_sst.nc")
print(ds.data_vars)
# Data variables:
#     time_bnds       (time, bnds) object 1870-01-01 17:59:59.588012 ... 2022-0...
#     sst             (time, latitude, longitude) float32 ...
#     longitude_bnds  (longitude, bnds) float32 -180.0 -179.0 ... 179.0 180.0
#     latitude_bnds   (latitude, bnds) float32 90.0 89.0 89.0 ... -89.0 -90.0

ds = xcdat.swap_lon_axis(ds, (0, 360))
print(ds.data_vars)
# Data variables:
#     time_bnds       (time, bnds) object 1870-01-01 17:59:59.588012 ... 2022-0...
#     latitude_bnds   (latitude, bnds) float32 90.0 89.0 89.0 ... -89.0 -90.0
#     longitude_bnds  (longitude, bnds) float32 0.0 1.0 1.0 ... 0.0 359.0 360.0
#     sst             (time, latitude, longitude) float32 -1e+03 -1e+03 ... nan
```

**Reply by mzelinka** [reply]
Looks fixed on my end too. Thanks!

**Reply by mzelinka** [reply]
Using xcdat v0.4.0, I just got an error when trying to to swap the longitude axes. I wonder if the issue is that coord_keys = get_dim_coords(ds, "X").coords.keys() on [L236](https://github.com/xCDAT/xcdat/blob/main/xcdat/axis.py#L236) of swap_lon_axis() returns not just the coordinate key for the "X" dimension but also "height" since this is a 2-meter air temperature.  

Manually choosing only the longitude key on [L260](https://github.com/xCDAT/xcdat/blob/main/xcdat/axis.py#L260) resolves the error. That is: ds= ds.sortby(list(coord_keys)[0], ascending=True) does not throw this error.  

See screen shots below.
![Screen Shot 2022-11-10 at 4 42 09 PM](https://user-images.githubusercontent.com/11380489/201235679-9bd38505-71f2-4be6-ac11-83946355873a.png)
![Screen Shot 2022-11-10 at 4 42 42 PM](https://user-images.githubusercontent.com/11380489/201235763-10481ecd-0546-457e-b50b-69e601bc68b0.png)

**Reply by pochedls** [reply]
Just logging some notes...this doesn't seem to happen with all datasets (e.g., this dataset doesn't seem to have an issue`/p/css03/esgf_publish/CMIP6/CMIP/NCAR/CESM2/amip/r1i1p1f1/Amon/tas/gn/v20190218/tas_Amon_CESM2_amip_r1i1p1f1_gn_195001-201412.nc`). 

Is this a one-off or happening to a lot of daily datasets? 

Maybe this should be logged as a bug.

**Reply by mzelinka** [reply]
Thanks @pochedls . My example may be an edge case.  I can confirm that for the file you listed, the get_dim_coords(ds, "X").coords.keys() command does not return height, just lon, as expected / desired (see screen shot).
![Screen Shot 2022-11-11 at 9 12 08 AM](https://user-images.githubusercontent.com/11380489/201393553-72ebeefd-59fc-4c44-9dad-663bc8e2d3c2.png)

**Reply by tomvothecoder** [reply]
> Using xcdat v0.4.0, I just got an error when trying to to swap the longitude axes. I wonder if the issue is that coord_keys = get_dim_coords(ds, "X").coords.keys() on [L236](https://github.com/xCDAT/xcdat/blob/main/xcdat/axis.py#L236) of swap_lon_axis() returns not just the coordinate key for the "X" dimension but also "height" since this is a 2-meter air temperature.

Hey @mzelinka and @pochedls, this issue is related to  `height` being a singleton coordinate. Singleton coordinate variables are  included alongside other coordinate variables (refer to code example below). There is an open xarray issue regarding singletons here: https://github.com/pydata/xarray/issues/6196

```python
import xcdat as xc

fnames = "/p/css03/esgf_publish/CMIP6/CMIP/BCC/BCC-CSM2-MR/historical/r1i1p1f1/day/tasmax/gn/v20181126/*.nc"

ds = xc.open_mfdataset(fnames)

print(xc.get_dim_coords(ds, axis="X"))

# Notice how `height` is included in the `lon` coordinate var.

# <xarray.DataArray 'lon' (lon: 320)>
# array([  0.   ,   1.125,   2.25 , ..., 356.625, 357.75 , 358.875])
# Coordinates:
#   * lon      (lon) float64 0.0 1.125 2.25 3.375 4.5 ... 355.5 356.6 357.8 358.9
#     height   float64 2.0
# Attributes:
#     bounds:         lon_bnds
#     units:          degrees_east
#     axis:           X
#     long_name:      Longitude
#     standard_name:  longitude
```

### How it affects `swap_lon_axis()`
In `swap_lon_axis()`, we are getting all of the coordinates keys using `lon`, which includes `height`.
This causes `swap_lon_axis(ds, sort_ascending=True)` to break on this file.
https://github.com/xCDAT/xcdat/blob/d5d4fd77cbe0e23cc6452c2a1d421923d800b31b/xcdat/axis.py#L235-L236
https://github.com/xCDAT/xcdat/blob/d5d4fd77cbe0e23cc6452c2a1d421923d800b31b/xcdat/axis.py#L259-L260

### Temporary workaround: 
* Set `sort_ascending=False` in `swap_lon_axis()` and perform sorting manually

### The fix for `swap_lon_axis()` is to ignore singleton coordinate keys.

I will open a new issue to fix this.

**Comment by tomvothecoder** [comment]
I am going to lock this thread because v0.3.0 is outdated. Please refer to #391 for v0.4.0 testing feedback.
